src/MergeContext.js

```diff
@@ -454,6 +454,252 @@ class BranchPosition
+        const label = this._paragraphLabel(str);
+        return (label && label.includes('-')) ? label : null;
+    }
+
+    // basic checks for an unparsed message
+    _checkRaw(message) {
+        // CRs in CRLF sequences already removed
+        // other CRs are not treated specially (and are banned)
+        const lines = message.split('\n');
+        for (let i = 0; i < lines.length; ++i) {
+            const line = lines[i];
+            const invalidPosition = this._invalidCharacterPosition(line);
+
+            if (invalidPosition !== -1)
+                throw new Error(`bad character at line ${i}, offset ${invalidPosition}`);
+
+            // allow excessively long whitespace-only lines
+            // that some copy-pasted PR descriptions may include
+            const trimmedLine = this._trim(line);
+            if (trimmedLine.length > 72)
+                throw new Error(`too long line '${trimmedLine}'`);
+        }
+    }
+
+    _parse(prDescriptionRaw) {
+        const prDescription = prDescriptionRaw.replace(/\r+\n/g, '\n');
+        this._checkRaw(prDescription);
+        const prDescriptionWithoutHeader = this._extractHeader(prDescription);
+        const prDescriptionWithoutHeaderAndTrailer = this._extractTrailer(prDescriptionWithoutHeader);
+        this._parseBody(prDescriptionWithoutHeaderAndTrailer);
+    }
+
+    // authorField is a {name, value, raw}
+    _parseAuthor(authorField) {
+        const trimmedValue = this._trim(authorField.value);
+        const cred = trimmedValue.match(/^ ([\w][^@<>,]*) <(\S+@\S+\.\S+)>$/);
+        if (!cred)
+            throw new Error(`unsupported ${authorField.name} value format: ${authorField.value}`);
+
+        if (cred[0].includes(','))
+            throw new Error(`${authorField.name} author name with a comma: ${authorField.value}`);
+
+        return {name: cred[1].trim(), email: cred[2].trim()};
+    }
+
+    _extractHeader(prDescriptionRaw) {
+        const prDescription = this._trim(prDescriptionRaw);
+        if (this._startsWithFieldName(prDescription) === 'Authored-by') {
+            let tokenizer = new FieldsTokenizer(prDescription);
+            const authorField = tokenizer.nextField();
+            assert(authorField);
+            this._customAuthor = this._parseAuthor(authorField);
+            if (!tokenizer.atEnd())
+                throw new Error(`unexpected header lines after a single Authored-by attribute`);
+            return tokenizer.remaining();
+        } else {
+            return prDescription;
+        }
+    }
+
+    _extractTrailer(prDescriptionWithoutHeaderRaw) {
+        const prDescriptionWithoutHeader = this._trim(prDescriptionWithoutHeaderRaw);
+        // index of the last occurrence of '\n\n'
+        let trailerCandidateIndex = prDescriptionWithoutHeader.search(/\n\n(?!.*\n\n)/s);
+        // Does the text have at most one paragraph?
+        trailerCandidateIndex = trailerCandidateIndex < 0 ? 0 : trailerCandidateIndex + 2
```

Comment:
Missing semicolon, but I would also add parens to avoid any doubts regarding the `?` RHS expression scope.

```suggestion
        trailerCandidateIndex = trailerCandidateIndex < 0 ? 0 : (trailerCandidateIndex + 2);
```

Reply:
Ok.